File: jdollarx/src/main/java/org/lab49/jdollarx/ElementProperty.java
package org.lab49.jdollarx;


public interface ElementProperty {
    String toXpath() ;

    default public ElementProperty or(ElementProperty p) {
        return new ElementProperties.Or(this, p);
    }

    default public ElementProperty and(ElementProperty prop) {
        return new ElementProperties.And(this, prop);
    }

    default public ElementProperty andNot(ElementProperty prop) { return and(new ElementProperties.Not(prop)); }
}

File: dollarx/src/main/java/com/github/loyada/dollarx/ElementProperty.scala
package com.github.loyada.dollarx

/**
  * Created by danny on 2/6/16.
  */
trait ElementProperty {

    def toXpath: String

    def or(p: ElementProperty): ElementProperty = {
      return new ElementPropertiesHelper.Or(this, p)
    }

    def and(prop: ElementProperty): ElementProperty = {
      return new ElementPropertiesHelper.And(this, prop)
    }

    def andNot(prop: ElementProperty): ElementProperty = {
      return and(new ElementPropertiesHelper.Not(prop))
    }
}
